---
layout: post
title:  "Iterable 과 Iterator"
author: "강신일"
comments: False
tags: python
---
python 을 배우면서 검색을 통해 알게되었던 내용들을 정리하는 글입니다.<br>

### Iterable

`Iterable` 은 Python 에서 데이터스트림의 내부요소를 한번에 하나씩 반환할 수 있는 객체를 뜻한다.
`Iterable` 하다고 정의할수 있는 예로는 모든 Sequence Type(list, tuple, str) 과 Dictionary, File 혹은 `__iter__` 또는 `__getitem__` 함수가 정의되어 있는 클래스이다.

```
# check iterable object
>>> import collections
>>> iter_list = [1,2,3,4,5]
>>> isinstance(iter_list, collections.Iterable)
True
>>> iter_tuple = (1,2,3,4,5)
>>> isinstance(iter_tuple, collections.Iterable)
True
>>> iter_set = {1,2,3,4,5}
>>> isinstance(iter_set, collections.Iterable)
True
>>> iter_dictionary = {"a":1, "b": 2, "c": 3, "d":4, "e":5}
>>> isinstance(iter_dictionary, collections.Iterable)
True
>>> iter_str = "12345"
>>> isinstance(iter_str, collections.Iterable)
True
>>> iter_range = range(5)
>>> isinstance(iter_range, collections.Iterable)
True
>>> non_iter_int = 12345
>>> isinstance(non_iter_int, collections.Iterable)
False
>>> non_iter_boolean = False
>>> isinstance(non_iter_boolean, collections.Iterable)
False
```

`Iterable` 한 객체를 내장함수 `iter()` 의 argument 로 전달하면 `iter()` 는 `iterator` 객체를 생성해서 반환한다.
```
type(iter([1,2,3,4,5])) # <class 'list_iterator'>
```

### Iterator
`Iterator` 는 데이터 스트림을 나타내며 `next`(Python2) 또는 `__next__` 함수가 정의되어 있는 객체이다.
`Iterator` 는 내장함수 `next()` 함수를 반복적으로 호출해서 데이터스트림의 내부요소에 접근할 수 있다. 만약 `next()` 로 호출할 내부요소가 더이상 없다면 `StopIteration Exception` 을 발생시킨다. 또한 `iterator` 는 `__iter__` 함수를 가지고 있어서 하므로 `Iterator` 는 `Iterable` 하다.

```
>>> list_iterator = iter([1,2,3,4,5])
>>> next(list_iterator)
1
>>> next(list_iterator)
2
>>> list_iterator.__next__()
3
>>> list_iterator.__next__()
4
>>> list_iterator.__next__()
5
>>> list_iterator.__next__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

우리가 사용하는 [for](https://docs.python.org/3/reference/compound_stmts.html#for) 문도 다음과 같이 `iterable` 을 이용해서 내부의 값을 하나씩 가져오고 있다.
```
for element in iterable:
```

for 문에서는 `iterable` 을 사용하지만 우리는 `iter()`, `next()` 를 사용하지 않고 `StopIteration Exception` 을 의식하지 않고 사용하고 있다.
우리가 이렇게 사용할 수 있는 이유는 for 문으로 반복하는 동안 Python 내부에서는 임시로 `iterator` 를 생성하고 `next()` 를 호출해주며 `StopIteration Exception` 을 핸들링 해주기 때문이다.
for 문의 내부 구성은 다음과 같은 방식으로 동작한다고 이해하면 좋다.

```
it = iter(some_primes)
while True:
    try:
        prime = next(it)
    except StopIteration:
        break
    else:
        print(prime)
```

### 나만의 Iterator 만들기

이번에는 위에서 배운 내용들을 바탕으로 나만의 반복문 `iterator` 를 만들어보자. <br>
밑의 예제는 `next()` 가 호출될때마다 max 보다 작을때까지 2씩 증가하는 `iterator` 이다.
```
class MyCounter:
    def __init__(self, max):
        self.current = 0
        self.max = max

    def __iter__(self):
        return self

    def __next__(self):
        if self.current < self.max:
            temp_current = self.current
            self.current += 2
            return temp_current        
        else:                          
            raise StopIteration  

>>> for i in MyCounter(5):
        print(i)

0
2
4

>>>  isinstance(MyCounter(5), collections.Iterable)
True
```

참고
1. [https://docs.python.org/2/glossary.html#term-iterable](https://docs.python.org/2/glossary.html#term-iterable)
2. [https://bluese05.tistory.com/55](https://bluese05.tistory.com/55)
3. [https://dojang.io/mod/page/view.php?id=2405](https://dojang.io/mod/page/view.php?id=2405)
